pkg/controllers/mcs/mcs_controller.go

```diff
@@ -0,0 +1,387 @@
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+		return err
+	}
+	return c.deriveServiceImportFromMultiClusterService(ctx, mcs, ownerSvc)
+}
+
+func (c *MultiClusterServiceController) newMultiClusterServiceFunc() handler.MapFunc {
+	return func(object client.Object) (requests []reconcile.Request) {
+		mcs := &networkingv1alpha1.MultiClusterService{}
+		err := c.Client.Get(context.TODO(), types.NamespacedName{Namespace: object.GetNamespace(), Name: object.GetName()}, mcs)
+		if err != nil {
+			return nil
+		}
+		requests = append(requests, reconcile.Request{NamespacedName: types.NamespacedName{
+			Namespace: mcs.Namespace,
+			Name:      mcs.Name,
+		}})
+		return
+	}
+}
+
+// SetupWithManager creates a controller and register to controller manager.
+func (c *MultiClusterServiceController) SetupWithManager(mgr controllerruntime.Manager) error {
+	return controllerruntime.NewControllerManagedBy(mgr).
+		For(&networkingv1alpha1.MultiClusterService{}).
+		Watches(&source.Kind{Type: &corev1.Service{}},
+			handler.EnqueueRequestsFromMapFunc(c.newMultiClusterServiceFunc()),
+			builder.WithPredicates(predicate.Funcs{
+				CreateFunc: func(createEvent event.CreateEvent) bool { return true },
+				DeleteFunc: func(deleteEvent event.DeleteEvent) bool { return false },
+				UpdateFunc: func(updateEvent event.UpdateEvent) bool { return false },
+			})).
+		Watches(&source.Kind{Type: &mcsv1alpha1.ServiceImport{}},
+			handler.EnqueueRequestsFromMapFunc(c.newMultiClusterServiceFunc()),
+			builder.WithPredicates(predicate.Funcs{
+				CreateFunc: func(createEvent event.CreateEvent) bool { return false },
+				DeleteFunc: func(deleteEvent event.DeleteEvent) bool { return true },
+				UpdateFunc: func(updateEvent event.UpdateEvent) bool { return false },
+			})).
+		Watches(&source.Kind{Type: &mcsv1alpha1.ServiceExport{}},
+			handler.EnqueueRequestsFromMapFunc(c.newMultiClusterServiceFunc()),
+			builder.WithPredicates(predicate.Funcs{
+				CreateFunc: func(createEvent event.CreateEvent) bool { return false },
+				DeleteFunc: func(deleteEvent event.DeleteEvent) bool { return true },
+				UpdateFunc: func(updateEvent event.UpdateEvent) bool { return false },
+			})).
+		WithOptions(controller.Options{
+			RateLimiter:             ratelimiterflag.DefaultControllerRateLimiter(c.RateLimiterOptions),
+			MaxConcurrentReconciles: 3,
+		}).
+		Complete(c)
+}
+
+// removeFinalizer remove finalizer from the given MultiClusterService
+func (c *MultiClusterServiceController) removeFinalizer(obj client.Object) (controllerruntime.Result, error) {
+	if !controllerutil.ContainsFinalizer(obj, MultiClusterServiceControllerFinalizer) {
+		return controllerruntime.Result{}, nil
+	}
+
+	controllerutil.RemoveFinalizer(obj, MultiClusterServiceControllerFinalizer)
+	err := c.Client.Update(context.TODO(), obj)
+	if err != nil {
+		return controllerruntime.Result{Requeue: true}, err
+	}
+	return controllerruntime.Result{}, nil
```

Comment:
I see that currently all controllers are using this method. Regarding your suggestion, I believe it can be modified uniformly with another pull request.